builtins/web/event/event-target.cpp

```diff
@@ -0,0 +1,557 @@
+  // against event's related target always returns self.
+  if (!invoke_listeners(cx, target, event)) {
+    return false;
+  }
+
+  // 7. Set event's eventPhase attribute to NONE.
+  // 8. Set event's currentTarget attribute to null.
+  // 9. Set event's path to the empty list.
+  // 10. Unset event's dispatch flag, stop propagation flag, and stop immediate propagation flag.
+  Event::set_phase(event, Event::Phase::NONE);
+  Event::set_current_target(event, nullptr);
+  Event::set_dispatched(event, false);
+  Event::set_stop_propagation(event, false);
+  Event::set_stop_immediate_propagation(event, false);
+
+  // 11. If clearTargets is true:
+  Event::set_related_target(event, nullptr);
+  // 12. If activationTarget is non-null:
+  // N/A
+  // 13. Return false if event's canceled flag is set; otherwise true.
+  rval.setBoolean(!Event::is_default_prevented(event));
+  return true;
+}
+
+// https://dom.spec.whatwg.org/#concept-event-listener-invoke
+bool EventTarget::invoke_listeners(JSContext *cx, HandleObject target, HandleObject event) {
+  MOZ_ASSERT(is_instance(target));
+
+  // 1. Set event's target to the shadow-adjusted target of the last struct in event's path,
+  //   that is either struct or preceding struct, whose shadow-adjusted target is non-null.
+  // 2. Set event's relatedTarget to struct's relatedTarget.
+  // 3. Set event's touch target list to struct's touch target list.
+  // We only use a single target here as it would appear in a Even#path[0];
+  // - shadow adjusted target == target
+  // - relatedTarget == target
+  Event::set_phase(event, Event::Phase::AT_TARGET);
+  Event::set_target(event, target);
+  Event::set_related_target(event, target);
+
+  // 4. If event's stop propagation flag is set, then return.
+  if (Event::is_stopped(event)) {
+    return true;
+  }
+
+  // 5. Initialize event's currentTarget attribute to struct's invocation target.
+  Event::set_current_target(event, target);
+  // 6. Let listeners be a clone of event's currentTarget attribute value's event listener list.
+  auto list = listeners(target);
+  JS::RootedVector<EventListener> list_clone(cx);
+  for (auto &listener : *list) {
+    if (!list_clone.append(listener)) {
+      return false;
+    }
+  }
+
+  // 7. Let invocationTargetInShadowTree be struct's invocation-target-in-shadow-tree.
+  // N/A
+  // 8. Let found be the result of running inner invoke with event, listeners, phase,
+  auto found = false;
+  if (!inner_invoke(cx, event, list_clone, &found)) {
+    return false;
+  }
+
+  // 9. If found is false and event's isTrusted attribute is true:
+  // N/A
+  return true;
+}
+
+
+// https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke
+bool EventTarget::inner_invoke(JSContext *cx, HandleObject event, JS::HandleVector<EventListener> list, bool *found) {
+  // 1. Let found be false.
+  *found = false;
+
+  // 2. For each listener of listeners, whose removed is false:
+  for (auto &listener : list) {
```

Comment:
If I read things correctly, then exceptions thrown while invoking any individual listener shouldn't abort the loop, but instead keep it running. The easiest way to do that might be to move the entire loop body into its own function, except for the substeps of step 11; the error reporting.

I'm a bit surprised that this doesn't seem to be covered by WPT tests. Unless my reading is wrong, and aborting is correct here. Can you maybe verify this with a new test that you also (manually, in browser devtools) run in a browser?

Reply:
Indeed the loop should not be aborted. I've confirmed this in a browser and added integration test.